z3c/dependencychecker/db.py

```diff
@@ -43,3 +48,186 @@ def _filter_duplicates(self, imports):
+
+    def get_missing_test_imports(self):
+        filters = (
+            self._filter_known_packages,
+            self._filter_only_testing_imports,
+            self._filter_own_package,
+            self._filter_requirements,
+            self._filter_test_requirements,
+            self._filter_python_standard_library,
+        )
+        filtered = self._process_pipeline(self.imports_used, filters)
+        unique_imports = self._get_unique_imports(imports_list=filtered)
+        return unique_imports
+
+    def get_unneeded_requirements(self):
+        all_but_test_requirements = self._requirements.copy()
+        for extra in self._extras_requirements:
+            all_but_test_requirements.update(self._extras_requirements[extra])
+        test_requirements = self._get_test_extra()
+        if test_requirements:
+            for dotted_name in test_requirements:
+                all_but_test_requirements.remove(dotted_name)
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(all_but_test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def requirements_that_should_be_test_requirements(self):
+        filters = (
+            self._filter_testing_imports,
+            self._filter_python_standard_library,
+            self._filter_own_package,
+        )
+        non_testing_imports = self._process_pipeline(
+            self.imports_used,
+            filters,
+        )
+        requirements_not_used = [
+            x
+            for x in self._requirements
+            if self._filter_requirements_lists(non_testing_imports, x)
+        ]
+        filters = (
+            self._filter_only_testing_imports,
+            self._filter_python_standard_library,
+            self._filter_own_package,
+        )
+        testing_imports = self._process_pipeline(
+            self.imports_used,
+            filters,
+        )
+        should_be_test_requirements = [
+            x
+            for x in requirements_not_used
+            if not self._filter_requirements_lists(testing_imports, x)
+        ]
+        unique = self._get_unique_imports(
+            imports_list=should_be_test_requirements,
+        )
+        return unique
+
+    def get_unneeded_test_requirements(self):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return []
+
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def _process_pipeline(self, objects, filters):
```

Comment:
This needs a quick explanation of what you're doing here.
A filter function should apparently return False or True. What is the meaning when it returns "True"? Does the item gets filtered out or does it get included?

Reply:
Done